api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,197 @@
+	MachineDrainRuleDrainBehaviorDrain MachineDrainRuleDrainBehavior = "Drain"
+
+	// MachineDrainRuleDrainBehaviorSkip means the drain for a Pod should be skipped.
+	MachineDrainRuleDrainBehaviorSkip MachineDrainRuleDrainBehavior = "Skip"
+)
+
+// MachineDrainRuleSpec defines the spec of a MachineDrainRule.
+type MachineDrainRuleSpec struct {
+	// drain configures if and how Pods are drained.
+	// +required
+	Drain MachineDrainRuleDrainConfig `json:"drain"`
+
+	// machines defines to which Machines this MachineDrainRule should be applied.
+	//
+	// If machines is empty, the MachineDrainRule applies to all Machines.
+	// If machines contains multiple selectors, the results are ORed.
+	// Within a single selector the results of selector and clusterSelector are ANDed.
+	// This field is NOT optional.
+	//
+	// Example: Selects control plane Machines in all Clusters and
+	//          Machines with label "os" == "linux" in Clusters with label
+	//          "stage" == "production".
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: cluster.x-k8s.io/control-plane
+	//        operator: Exists
+	//  - selector:
+	//      matchLabels:
+	//        os: linux
+	//    clusterSelector:
+	//      matchExpressions:
+	//      - key: stage
+	//        operator: In
+	//        values:
+	//        - production
+	//
+	// +required
+	// +kubebuilder:validation:MaxItems=32
+	Machines []MachineDrainRuleMachineSelector `json:"machines"`
+
+	// pods defines to which Pods this MachineDrainRule should be applied.
+	//
+	// If pods is empty, the MachineDrainRule applies to all Pods.
+	// If pods contains multiple selectors, the results are ORed.
+	// Within a single selector the results of selector and namespaceSelector are ANDed.
+	// This field is NOT optional.
+	//
+	// Example: Selects Pods with label "app" == "logging" in all Namespaces and
+	//          Pods with label "app" == "prometheus" in the "monitoring"
+	//          Namespace.
+	//
+	//  - selector:
+	//      matchExpressions:
+	//      - key: app
+	//        operator: In
+	//        values:
+	//        - logging
+	//  - selector:
+	//      matchLabels:
+	//        app: prometheus
+	//    namespaceSelector:
+	//      matchLabels:
+	//        kubernetes.io/metadata.name: monitoring
+	//
+	// +required
+	// +kubebuilder:validation:MaxItems=32
+	Pods []MachineDrainRulePodSelector `json:"pods"`
+}
+
+// MachineDrainRuleDrainConfig configures if and how Pods are drained.
+type MachineDrainRuleDrainConfig struct {
+	// behavior defines the drain behavior.
+	// Can be either "Drain" or "Skip".
+	// +required
+	Behavior MachineDrainRuleDrainBehavior `json:"behavior"`
+
+	// order defines the order in which Pods are drained.
+	// Pods with higher order are drained after Pods with lower order.
+	// order can only be set if behavior is set to "Drain".
+	// If order is not set, 0 will be used.
+	// +optional
+	Order *int32 `json:"order,omitempty"`
```

Comment:
Can this be negative?

Reply:
>// If order is not set, 0 will be used.

Any reason to prefer implementing the above with a pointer rather than let it be an int and default to their zero value?

Reply:
> Can this be negative?

Yes. Similar to PriorityClass.value

Reply:
> Any reason to prefer implementing the above with a pointer rather than let it be an int and default to their zero value?

Basically I have the following requirements:
* 0 should be default
* if 0 is set, it should be visible (so using int32 & omitempty is not an option)

Option 1.
```go
	// +optional
	Order *int32 `json:"order,omitempty"`
```

Option 2.
```go
	// +optional
	Order int32 `json:"order"`
```

Option 1 makes it possible to tell if the user intentionally set 0. But I'm also okay with Option 2, if we prefer.

@enxebre @JoelSpeed WDYT?

Reply:
When would we need to distinguish between the user setting the value to zero, vs it being the default?

Do you ever see a future in which zero is not the default behaviour?

I'm leaning towards option 2, but we should perhaps consider the behaviour. If a user writes yaml without this value, and then applies it, the value will not be set to 0 without a `default` set in the spec, when read by the user, so perhaps we need that too?

Does having the value marshalled by a controller and written out cause conflict with SSA perhaps?

Reply:
Per API conventions, it should probably be option 1, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Reply:
Given everything we've said above, I think then we can go non-pointer, but should make sure the defaulting is added to the schema
```
// +default=0
// +optional
Order int32 `json:"order"`

```

Reply:
This decision (if we are agreed) isn't applied yet

Reply:
Yup, will be done on the next commit

Reply:
Hm. I think this doesn't work. The problem is how can we then not set the field if the behavior is Skip?

Reply:
With that. Maybe Option 1 is the only option that works?

Reply:
Ahh yes, good point, since this isn't allowed to be set with skip, it must be `omitempty` and not have a default, agreed, it must stay as option one to fit the validation for the skip case

Technically, this is creating a discriminated union right? Some options only valid when a certain field is set to a certain value from an enum. If we followed the DU pattern to the letter this would be easier to understand and easier to validate

Reply:
> Technically, this is creating a discriminated union right? Some options only valid when a certain field is set to a certain value from an enum. If we followed the DU pattern to the letter this would be easier to understand and easier to validate

Not sure. At this point we only don't want to allow to set order with Skip, because that specific case doesn't make sense.

I would imagine that for potential additional future values of behavior (other kinds of Drain) order always should be set (e.g. let's say we add some sort of ForceDrain / MustDrain)

Reply:
Even in DUs where there are fields that would be shared between multiple values, the correct structure of this future DU would be
```
drain:
  mode: Skip | Drain | ForceDrain
  drain:
    order: 0
  forceDrain:
    order: 0
```

Reply:
@vincepri @fabriziopandini What do you think?

Reply:
Do we expect there may be further drain related configuration in the future?

Reply:
Could we keep default to `nil` and only allow positive values between 1 -> whatever?

Reply:
If we only allow positive values. That means that nobody will be able to drain earlier than the Pods without any specific order configured (which have 0)

Reply:
> Do we expect there may be further drain related configuration in the future?

I don't have anything in mind at the moment

Reply:
IMO it should be *int32 (+optional): 
- With behaviour skip the value must not be set(nil), and it must not show up in the yaml (omit empty)
- With behaviour drian the value must be set, it could be <0 (before), 0 (with the other pods), >0 (after pods)

I will also keep the API it is and unblock the work, given that
a) all the use case we have in mind are covered by this API, and we don't much to speculate what could go on top 
b) if some new requirement pops up we can shuffle things around in the next release when we cut v1beta2

Also, time is running quickly for this release...

Reply:
> a) applies to all modes/behaviors (I would assume we duplicate into every "mode field"?)
b) has nothing to do with the mode/behavior?

DUs often aren't pretty, but they are very declarative. A, yes, goes into every mode sub struct, b, should be outside of the DU.

You're right though, at the moment, this is pretty basic and likely not too much an issue since we could revise this during a transition to v1

Reply:
Thx for the discussion everyone! So I'll keep it as is